File: src-tauri/src/commands/local_model.rs
use tauri::{command, State};

use crate::{contexts::local_model::ModelInfo, lifecycle::app_state::AppState};

trait ResultExt<T> {
    fn context(self, msg: &str) -> Result<T, String>;
}

impl<T, E: std::fmt::Display> ResultExt<T> for Result<T, E> {
    fn context(self, msg: &str) -> Result<T, String> {
        self.map_err(|e| format!("{}: {}", msg, e))
    }
}

#[command]
#[specta::specta]
pub async fn list_local_models(app_state: State<'_, AppState>) -> Result<Vec<ModelInfo>, String> {
    app_state
        .local_model_context
        .list_models()
        .await
        .context("Failed to list local models")
}

#[command]
#[specta::specta]
pub async fn download_local_model(
    model_id: String,
    app_state: State<'_, AppState>,
) -> Result<(), String> {
    app_state
        .local_model_context
        .download_model(&model_id)
        .await
        .context("Failed to start model download")
}

#[command]
#[specta::specta]
pub async fn cancel_model_download(
    model_id: String,
    app_state: State<'_, AppState>,
) -> Result<(), String> {
    app_state.local_model_context.cancel_download(&model_id);
    Ok(())
}

#[command]
#[specta::specta]
pub async fn delete_local_model(
    model_id: String,
    app_state: State<'_, AppState>,
) -> Result<(), String> {
    app_state
        .local_model_context
        .delete_model(&model_id)
        .await
        .context("Failed to delete model")
}

#[command]
#[specta::specta]
pub async fn get_active_local_model(
    app_state: State<'_, AppState>,
) -> Result<Option<String>, String> {
    app_state
        .local_model_context
        .get_active_model_id()
        .await
        .context("Failed to get active model")
}

#[command]
#[specta::specta]
pub async fn set_active_local_model(
    model_id: String,
    app_state: State<'_, AppState>,
) -> Result<(), String> {
    app_state
        .local_model_context
        .set_active_model(&model_id)
        .await
        .context("Failed to set active model")
}

#[command]
#[specta::specta]
pub async fn generate_local_insights(
    request: crate::service::GeminiRequest,
    app_state: State<'_, AppState>,
) -> Result<String, String> {
    app_state
        .local_model_context
        .generate_insights(&request)
        .await
        .context("Failed to generate local AI insights")
}

File: src/api/local-model.ts
import { commands, type ModelInfo } from "@/src/bindings";
import { Result } from "@/src/lib/result";

export async function listLocalModels(): Promise<Result<ModelInfo[], string>> {
  const res = await commands.listLocalModels();
  return res.status === "ok" ? Result.Ok(res.data) : Result.Err(res.error ?? "");
}

export async function downloadLocalModel(modelId: string): Promise<Result<null, string>> {
  const res = await commands.downloadLocalModel(modelId);
  return res.status === "ok" ? Result.Ok(res.data) : Result.Err(res.error ?? "");
}

export async function cancelModelDownload(modelId: string): Promise<Result<null, string>> {
  const res = await commands.cancelModelDownload(modelId);
  return res.status === "ok" ? Result.Ok(res.data) : Result.Err(res.error ?? "");
}

export async function deleteLocalModel(modelId: string): Promise<Result<null, string>> {
  const res = await commands.deleteLocalModel(modelId);
  return res.status === "ok" ? Result.Ok(res.data) : Result.Err(res.error ?? "");
}

export async function getActiveLocalModel(): Promise<Result<string | null, string>> {
  const res = await commands.getActiveLocalModel();
  return res.status === "ok" ? Result.Ok(res.data) : Result.Err(res.error ?? "");
}

export async function setActiveLocalModel(modelId: string): Promise<Result<null, string>> {
  const res = await commands.setActiveLocalModel(modelId);
  return res.status === "ok" ? Result.Ok(res.data) : Result.Err(res.error ?? "");
}
